### ISSUE TITLE
Clippy fix creates broken code with saturating sub

### ISSUE THREAD

**Issue by zSuperx** [rust-lang/rust-clippy#16307]
### Summary

I stumbled across this self-proclaimed bug while creating an application where the code contained a snippet resembling the given MRE.

Specifically, if `x` is explicitly typed, clippy suggests `100.saturating_sub(x)` (understandably so). However, it does not explicitly type `100` (via the `_u8` suffix) if `y` is not explicitly typed, which results in a compiler error post-fix. It makes sense that this is not a safe assumption to make, but does bring into question the initial clippy suggestion.

And as per the request in `cargo clippy`'s output, I am simply here to report this bug to the Rust team! However, I have not yet checked for an existing bug report of this class, so I apologize for any issue duplication.

In any case, thanks to the Rust team for all their hard work!

### Reproducer

Code:

```rust
fn main() {
    let x: u8 = 100;

    let y = if x >= 100 {
        0
    } else {
        100 - x
    };

    println!("{y}");
}
```

Current output:
```
❯ cargo clippy --fix --allow-dirty
    Checking clippy-bug v0.1.0 (/home/zsuper/fun/clippy-bug)
warning: failed to automatically apply fixes suggested by rustc to crate `clippy_bug`

after fixes were automatically applied the compiler reported errors within these files:

  * src/main.rs

This likely indicates a bug in either rustc or cargo itself,
and we would appreciate a bug report! You're likely to see
a number of compiler warnings after this message which cargo
attempted to fix but failed. If you could open an issue at
https://github.com/rust-lang/rust-clippy/issues
quoting the full output of this command we'd be very appreciative!
Note that you may be able to make some more progress in the near-term
fixing code with the `--broken-code` flag

The following errors were reported:
error[E0689]: can't call method `saturating_sub` on ambiguous numeric type `{integer}`
 --> src/main.rs:4:17
  |
4 |     let y = 100.saturating_sub(x);
  |                 ^^^^^^^^^^^^^^
  |
help: you must specify a concrete type for this numeric value, like `i32`
  |
4 |     let y = 100_i32.saturating_sub(x);
  |                ++++

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0689`.
Original diagnostics will follow.

warning: manual arithmetic check found
 --> src/main.rs:4:13
  |
4 |       let y = if x >= 100 {
  |  _____________^
5 | |         0
6 | |     } else {
7 | |         100 - x
8 | |     };
  | |_____^ help: replace it with: `100.saturating_sub(x)`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#implicit_saturating_sub
  = note: `#[warn(clippy::implicit_saturating_sub)]` on by default

warning: `clippy-bug` (bin "clippy-bug") generated 1 warning (run `cargo clippy --fix --bin "clippy-bug" -p clippy-bug` to apply 1 suggestion)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.14s
```

### Version

```text
rustc 1.94.0-nightly (e7d44143a 2025-12-24)
binary: rustc
commit-hash: e7d44143a12a526488e4f0c0d7ea8e62a4fe9354
commit-date: 2025-12-24
host: x86_64-unknown-linux-gnu
release: 1.94.0-nightly
LLVM version: 21.1.8
```
